### PR TITLE
create the 'pool' framework

### PR DESCRIPTION
**Pull request by ggouaillardet**
Refs open-mpi/ompi#4264

### CODE REVIEW

**Comment by ggouaillardet** [comment]
:bot:mellanox:retest

**Comment by ggouaillardet** [comment]
:bot:ompi:retest

**Comment by ggouaillardet** [comment]
@jsquyres @rhc54 can you please have a look at this PR ?

File: opal/mca/pool/base/pool_base_open.c [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * Globals
+ */
+opal_pool_base_component_2_0_0_t *opal_pool = NULL;
+
+
+/*
+ * Function for finding and opening either all MCA components, or the one
+ * that was specifically requested via a MCA parameter.
+ */
+static int opal_pool_base_open(mca_base_open_flag_t flags)
```
Comment (rhc54) [inline opal/mca/pool/base/pool_base_open.c]: Why did this require a custom component open routine?

Comment (ggouaillardet) [reply opal/mca/pool/base/pool_base_open.c]: i did quite some copy/paste from existing frameworks, so this could be unnecessary. i will double check that.

**Comment by ggouaillardet** [comment]
@rhc54 just to be clear, the goal of the ```pool``` framework is not to fix a bug, but to provide a way of reducing the memory footprint.

- The ```trivial``` component mimic the current behavior (the same string is ```strdup```'ed several times).
- The ```strings``` component uses ref-counted ```opal_string_t```, so there is one object per unique string (e.g. hostname) on each MPI task.
- The next step is to implement a ```sm``` component so there is only one instance of a given hostname on a given host (e.g. the instance is in shared memory, and is shared between the MPI tasks and mpirun/orted).


As far as i am concerned, my main concern is that some backport/adjustement will be required when PR from master (with the ```pool``` framework) to a release branch without the ```pool``` framework.

If you do not expect a significant reduction of the memory footprint via this framework, then i am fine dropping it.

**Comment by rhc54** [comment]
@ggouaillardet good point - I stand corrected 😄

**Comment by bosilca** [comment]
I share @rhc54 feeling, frankly it's scary by its own size. We started as a discussion on how to prevent a segfault, and it evolved into a ref counted string framework. Adding the ```sm``` framework is indeed a huge improvement, but then we are basically replicating the key-value shared memory framework from PMIx.

This might not necessarily be a bad thing, if we extend it to encompass all MPI-needed infromation that can be shared among local processes. If PMIx cannot provide us with a sharing mechanism allowing us to limit the memory used per process to store business card information about the other processes, maybe we should have our own...

**Comment by rhc54** [comment]
To be clear, PMIx _does_ provide that sharing mechanism. Others felt it would create too strong a dependency on PMIx, however, and so a duplicate one inside OMPI was desired.

Personally, I'm unconvinced - but I suppose we could add a pmix component to this new framework to cover that option?

**Comment by bosilca** [comment]
Can we at least analyze that framework and if we decide that it match our goals slurp it in OPAL ? I would assume this will not be a problem with PMIx folks.

**Comment by rhc54** [comment]
Let me explain a bit. PMIx doesn't have a similar framework. What it does have is the ability to return a pointer to storage when asked for data instead of a malloc'd copy of it. We can also refcount that storage, if necessary, to protect it - though that won't solve the case of someone calling "free" on a string.

What I had offered to do before was to simply modify the opal/pmix_get function so we fetch pointers to values instead of copies. We would then have to audit OMPI to ensure that those values were no longer "free'd", so it is still a bit of an effort.

**Comment by bosilca** [comment]
@rhc54 this sounds like a lighter approach.

There is a significant amount of information that is duplicated among all local MPI processes. If we are serious about reducing memory requirements for OMPI that information must be stored into a shared location. So we might want to use a sharing framework for more than just the hostname.

**Comment by ggouaillardet** [comment]
folks, let's not get distracted by the ```SIGSEGV``` issue.
it is fixed, and while investigating the issue, i found some room to reduce the memory footprint, and here we are. at this stage, the fixed bug is no more relevant.

note the ```pool/strings``` component reduces the memory footprint by doing what storage folks call deduplication. this is deduplication per process, and the ```sm``` module could do deduplication per node.
does ```PMIx``` do deduplication internally ?
we can uses this framework for virtually any null terminated string. but if we want to use it for any objects, and we still want to perform deduplication, then null terminated strings and binary objects should be handled separately.
```hwloc``` can use a shared memory segment to reduce the memory footprint, but it is mandatory that all processes (mpirun/orted and MPI tasks) ```mmap``` it at the very same address (so the pointers it might contain can be dereferenced). i do not know how big a limitation this is.

now, we cannot assume ```PMIx``` is built with the shared memory datastore.
i do not think this impacts the functionality, but this definitely impacts memory footprint.

If ```PMIx``` is using a shared memory datastore, then the ```pool/pmix``` component could be as simple
as a copy of the ```pool/trivial``` component, but the ```free``` function would be just a no-op.
(e.g. you do not have to worry about someone calling ```free()``` on a string)

my primary concerns with this approach (e.g. relying on ```PMIx```) are
- i do not think PMIx does deduplication
- we cannot use the ```pool``` framework before and after the ```pmix``` framework is opened/closed.

i did not investigate any of these yet, and i could be wrong and/or there could be no issue.

an other approach (but basically limited to ```proc_hostname```) would be to use an index (4 bytes integer seems enough to me) instead of a (likely 8 bytes) pointer, and then add an indirection to access the string hostname.

at this stage, i think the ```pool``` framework give us the flexibility to experiment various techniques in order to reduce the memory footprint. we might end up with one component outperforming the other,
and at that point, we could simply remove the framework and directly use the component.

**Comment by rhc54** [comment]
@ggouaillardet I honestly don't understand why you would say that PMIx doesn't de-duplicate the data. What we do is create one shared memory instance of all data on the node. We offer the option of providing pointers to that data as individual elements. So in the full use-case, there would only be one actual instance of any job-level or modex data on a node, and every process would have pointers to it.

I don't know how you can "de-duplicate" more than that 😄 

For direct-launch on systems that don't offer PMIx, then we have to duplicate the data anyway. Fortunately, the latest information is that these systems are going away in the not-too-distant future, at least on machines of any scale. So it might not be a bad option, though I still acknowledge the validity of concerns over dependencies.

**Comment by ggouaillardet** [comment]
@rhc54 currently, if you are running ```n``` tasks on one node, then on a given MPI task and at the MPI level
```c
OPAL_MODEX_RECV_VALUE_OPTIONAL(ret, OPAL_PMIX_HOSTNAME, &proc->super.proc_name,
                                   &hostname, OPAL_STRING);
```
invoked on all the ```ompi_proc_t``` might lead to the same hostname being ```strdup```'ed ```n``` times.
with the ```pool/strings``` component, there would be only one (refcounted) string (e.g. deduplication)

in this scenario (e.g. ```n``` tasks running on a given node), how many strings are there for the hostname inside (the shared memory datastore of) ```PMIx``` ?
if there is only one, then that could not be better, but if the same hostname is stored ```n``` times (e.g. once per task), then there is room for improvement (e.g. deduplication)

makes sense ?

**Comment by rhc54** [comment]
@ggouaillardet Let me repeat what I said earlier - PMIx _can_ simply return a pointer to the memory instead of copying it. This is _not_ what we currently do. PMIx only stores what OMPI/ORTE tells it to store. We currently do _not_ store a minimal footprint set of data - instead, we push more info than is required because we know we are going to pull it a certain way.

ORTE currently stores a regex for the names of all the nodes in the job. This is expanded inside PMIx to an array of hostnames for the job. We do _not_ store a hostname for every process as that blows out the footprint.

If we truly want to minimize footprint, we could ask PMIx to retain just the regex, and then use the PMIx resolve APIs to ask that it create a hostname string for a particular node and return a pointer to it. Thus, subsequent resolve or modex recv requests would return the pointer to that string.

In that way, we only create strings for the hosts that we need, instead of doing so for every node in the job. Yes, it will involve some code changes on both sides - and we'd have to ensure we retain compatibility with non-PMIx opal/pmix components.

To be clear: I'm not advocating any particular option here. I'm just trying to clear up some apparent confusion over what PMIx does and doesn't do, and explain the PMIx-based option. If people want togo with this pool framework, then that's fine - I'm just clarifying an alternative..

**Comment by rhc54** [comment]
We had planned to discuss this at today's telecon, but ran out of time.. @gpaulsen would you please put this on next week's agenda?

**Comment by gpaulsen** [comment]
I added WIP-DNM until we discuss on next week's telcon

**Comment by jsquyres** [comment]
@bosilca @rhc54 @ggouaillardet Is this PR going anywhere?  It's over a year old.

**Comment by rhc54** [comment]
PMIx is working on returning a pointer, as described. I personally am not a fan of ref-counting strings, but I leave that up to you folks.

**Comment by ggouaillardet** [comment]
This is going nowhere so let's simply drop that PR.

**Pull request closed without merging**